jax/api.py

```diff
@@ -605,12 +647,43 @@ def vmap(fun, in_axes=0, out_axes=0):
            "or a (nested) tuple of those types, got {} and {} respectively.")
     raise TypeError(msg.format(type(in_axes), type(out_axes)))
 
+  def _check_axis_sizes(tree, vals, dims):
+    try:
+      sizes, = {x.shape[d] for x, d in zip(vals, dims) if d is not None}
+    except ValueError:
+      msg = "vmap got inconsistent sizes for array axes to be mapped:\n{}"
+      # we switch the error message based on whether args is a tuple of arrays,
+      # in which case we can produce an error message based on argument indices,
+      # or if it has nested containers.
+      # TODO(mattjj,phawkins): add a way to inspect pytree kind more directly
+      if tree == tree_flatten((core.unit,) * tree.num_leaves)[1]:
```

Comment:
lol I have no idea what this is doing, I trust your comment though

Reply:
It's checking that the treedef represents a flat tuple (by creating a flat tuple and forming its treedef and comparing for equality, since equality is the only access we have to treedefs aiui).